code/modules/storage/storage.dm

```diff
@@ -0,0 +1,422 @@
+/* ================================================================== */
+/* ------------------------- STORAGE DATUM -------------------------- */
+/* ================================================================== */
+
+// see storage.md for an intro to storage datums
+
+/// add storage to an atom
+/atom/proc/create_storage(storage_type, list/spawn_contents = list(), list/can_hold = list(), list/can_hold_exact = list(), list/prevent_holding = list(),
+		check_wclass = FALSE, max_wclass = W_CLASS_SMALL, slots = 7, sneaky = FALSE, opens_if_worn = FALSE)
+	var/list/previous_storage = list()
+	for (var/obj/item/I as anything in src.storage?.get_contents())
+		previous_storage += I
+	src.remove_storage()
+	src.storage = new storage_type(src, spawn_contents, can_hold, can_hold_exact, prevent_holding, check_wclass, max_wclass, slots, sneaky, opens_if_worn)
+	for (var/obj/item/I as anything in previous_storage)
+		src.storage.add_contents(I)
+
+/// remove atom's storage
+/atom/proc/remove_storage()
+	qdel(src.storage)
+	src.storage = null
+
+/// a datum for atoms that allows holdable storage of items in a hud
+/datum/storage
+	/// Types that can be held
+	var/list/can_hold = null
+	/// Exact types that can be held, in addition to can_hold, if it has types
+	var/list/can_hold_exact = null
+	/// Types that have a w_class holdable but that the storage will not hold
+	var/list/prevent_holding = null
+	/// If set, if can_hold is used, an item not in can_hold or can_hold_exact can fit in the storage if its weight is low enough
+	var/check_wclass = FALSE
+	/// Storage hud attached to the storage
+	var/datum/hud/storage/hud = null
+	/// Don't print a visible message on use
+	var/sneaky = FALSE
+	/// Prevent accessing storage when clicked when worn, ex. in pocket
+	var/opens_if_worn = FALSE
+	/// Maximum w_class that can be held
+	var/max_wclass = W_CLASS_SMALL
+	/// Number of storage slots, even numbers overlap the close button for the on-ground hud layout
+	var/slots = 7
+	/// Does moving the linked storage item cause anything to happen to stored items
+	var/move_triggered = TRUE
+	/// The storage item linked to this datum
+	var/atom/linked_item = null
+	/// All items stored
+	var/list/stored_items = null
+
+/datum/storage/New(atom/storage_item, list/spawn_contents, list/can_hold, list/prevent_holding, list/restricted_from_holding, check_wclass, max_wclass, \
+		slots, sneaky, opens_if_worn)
+	..()
+	src.stored_items = list()
+
+	src.linked_item = storage_item
+	src.hud = new (src)
+	src.can_hold = can_hold
+	src.can_hold_exact = can_hold_exact
+	src.prevent_holding = prevent_holding
+	src.check_wclass = check_wclass
+	src.max_wclass = max_wclass
+	src.slots = slots
+	src.sneaky = sneaky
+	src.opens_if_worn = opens_if_worn
+
+	RegisterSignal(src.linked_item, COMSIG_ITEM_DROPPED, .proc/storage_item_on_drop)
+
+	if (length(spawn_contents))
+		src.make_my_stuff(spawn_contents)
```

Comment:
Issue here: some stuff, like gang flyer cases, relies on the contents being created when the storage is first opened, instead of in New(). (for gang flyer cases, the `gang` variable is set after creating the box, so make_my_stuff happening off of `New()`, it causes the flyers to have `gang=null`)

Reply:
For the old code, it worked since make_my_stuff() was called through a SPAWN(), this removes that

I was able to fix gang flyers, and took another look through everything for /obj/item/storage make_my_stuff() that creates contents relying on a variable set before they're created, only other case that was an issue was satan briefcases which still worked since they used a SPAWN in make_my_stuff(). They were still working but I changed it up a bit to not use SPAWN().

Unless there are any other cases, I think this may be rare enough that it would be okay

Anything in the future, whoevers using it would need to make sure that created contents that rely on a variable to be set are added after storage creation